### PR TITLE
feat(qwen3_5): VLM support, and keep lm_head off the polar path

### PR DESCRIPTION
**Pull request by manjunathshiva**
Adds Qwen3.5-family VLM support (`model_type: qwen3_5`, e.g. **Qwen3.8-27B**) and fixes the one thing that stops those builds being usable.

No architecture work was needed — mlx-lm 0.31.3 and mlx-vlm 0.6.3 already carry `qwen3_5`, and `sanitize` drops the MTP block. Conversion takes ~18 s.

## The fix: `lm_head` stays off the polar path

Qwen3.8-27B has a **248,320-token vocabulary**, making `lm_head` 248320×5120 = **1.271B params** — the largest matrix in the model. `PolarQuantizedLinear` is fused only up to `_QMM_MAX_TOKENS` (256) tokens; above that it dequantizes the weight through full-size intermediates at ~14 B/param.

Measured on exactly that shape, 3-bit/g64:

| tokens | transient peak |
|---|---|
| 1 | +0.000 GiB |
| 256 | +0.116 GiB |
| **257** | **+13.141 GiB** |
| 2048 | +13.953 GiB |

The same shape as an 8-bit affine layer costs +1.895 GiB at 2048 tokens — the logits array itself, no weight-dequant term.

End to end, same build settings, same 1342-token prompt, only the skip differs:

| build | on disk | peak |
|---|---|---|
| `lm_head` affine (this PR) | 15.15 GiB | **22.85 GB** |
| `lm_head` polar (control) | 14.52 GiB | **33.34 GB** |

Polar buys **0.63 GB on disk** and costs **10.5 GB at runtime** — the difference between fitting a 36 GB Mac and not.

**Why it hides:** mlx-vlm's chunked prefill *discards* the chunk forward's return value, so MLX never evaluates the matmul. But chunking only engages **above** `prefill_step_size` (default 2048) — a prompt of **257–2048 tokens** takes the single-shot branch in `generate/ar.py`, which slices `logits[:, -1, :]` and therefore does evaluate `lm_head` across the whole sequence. That is an ordinary chat turn.

`muse_glimmer` already carried this skip for the same reason; `qwen3_5` did not.

## Validation

| build | size | peak (1.3K prompt) | decode | Opencode agentic |
|---|---|---|---|---|
| `tq4-g64` | 15.15 GiB | 22.85 GB | 11.4 tok/s | **pass** (6m12s) |
| `tq3-g64` | 12.88 GiB | 20.60 GB | 9.2 tok/s | pass (6m32s) |

Both fix a planted off-by-one end to end — run the given pytest command, read the file, make the minimal edit, re-run to green, explain with `file:line`. Vision passes on both (shapes, colours, positions, grounded bboxes). Native tool calls clean.

Predicted sizes matched actual **to the byte** on both builds.

`tq4` is *faster as well as better* than `tq3` here, consistent with the documented speed flip: the codebook path only earns its keep at 2-bit.

## Tests

4 new tests in `tests/test_qwen3_5_vlm.py`, verified to fail without the fix:

- `lm_head` is skipped for `qwen3_5`
- the skip is surgical — MLP, attention and GDN projections still quantize
- the vision tower still stays full precision
- the fused-kernel bound is a real step function (small shape, so it stays cheap)

Full suite: **551 passed, 5 skipped**, no regressions.

## Deliberately not changed

I suspected the Gated-DeltaNet gate projections (`in_proj_a` / `in_proj_b`, output dim 48, feeding `g = exp(-exp(A_log)·softplus(a + dt_bias))` — the per-step decay) needed protection, since they clear the `_should_quantize` `>= 32` guard by only 16. Two metrics said yes and both were wrong: a `∏g` ratio over 4096 steps is meaningless (`g ≈ 0.74`, so both products underflow to ~1e-536), and a mean-relative error was an artifact of a p1 tail at `g ≈ 0.002`.

Median relative error on `g` is **1.31% at 3-bit / 0.69% at 4-bit** with the memory horizon unchanged — ordinary quantization noise. Upstream mlx-vlm does not protect them either. No change made.

## Summary by CodeRabbit

* **New Features**
  * Added support and usage guidance for the Qwen3.8-27B model, including conversion, generation, benchmarks, and hardware requirements.
  * VLM conversion now includes required tokenizer, video-processing, and licensing files.
  * Added a prefill-step-size option to help control generation memory usage.

* **Bug Fixes**
  * Reduced temporary memory usage when processing Qwen3.5 VLMs with large vocabularies.
  * Preserved quantization for applicable model components while keeping the vision tower at full precision.

* **Documentation**
  * Updated architecture support details and documented memory behavior for Qwen3.5 and Qwen3.8 models.

### CODE REVIEW

**Comment by manjunathshiva** [comment]
Added a second, smaller fix while preparing the Qwen3.8-27B upload (021ae91).

`convert_vlm._AUX_FILES` was missing six files that belong beside the converted weights — I noticed because the converted directory had none of them:

- **`merges.txt` / `vocab.json`** — `tokenizer.json` alone covers the fast tokenizer, but anything constructing the slow one needs the BPE pair.
- **`video_preprocessor_config.json`** — video-capable VLMs (the Qwen3.5 family) keep a separate video preprocessor config.
- **`LICENSE` / `LICENSE.md` / `NOTICE`** — the licence should travel with derived weights rather than being re-added by hand at upload time.

On the video config I briefly thought I had introduced a regression: with the file present, `load_processor` complained that `Qwen3VLVideoProcessor` requires torchvision. That turned out to be an artifact of testing `load_processor` in isolation — the untouched `tq3` build (which never had the file) fails the same way, because the real path imports the model classes first and registers them. Verified end-to-end with `generate_vlm --image` both with and without the file: both work, and shipping it imposes no torchvision requirement on image-only use.

Suite still green: 551 passed, 5 skipped.

**Comment by manjunathshiva** [comment]
Worked through the review. Three of four were real and are fixed in c3dfdf6; one I'm declining, with reasoning.

## Fixed

**1. `--prefill-step-size` was not runnable on the VLM path** — the important one. `generate_vlm.py` had no such flag, so the recipe I documented would have died in argparse, and `turboquant-plan` recommends a value whenever the default won't fit. Added the flag and forwarded it to `mlx_vlm.generate` → `generate_step`.

Measured on Qwen3.8-27B tq4, 1342-token prompt:

| `--prefill-step-size` | peak | prefill |
|---|---|---|
| default (2048 — unchunked at this length) | 21.28 GB | 162 tok/s |
| **256** | **16.77 GB** | 77 tok/s |

−4.5 GB for 2.1× slower prefill. It worked via `turboquant-serve-vlm` already (that forwards to mlx-vlm's server, which has the flag); only the generate path was missing it.

**2. CodeQL — importing the value of a mutable attribute.** Correct, and it matters here rather than being pure style: `convert_vlm` swaps `_qm._should_quantize` for the duration of a conversion, so the module-level import would pin the pre-patch value. Now read off the module.

**3. Dependency versions.** Also correct, and worth more than a nitpick: the measurements were taken on **mlx-vlm 0.6.3**, which is *below* the `[vlm]` floor of `>=0.6.12` — so a fresh install would not have been the version I tested. Re-ran in a clean venv on **0.6.13**: identical memory peaks (21.28 / 16.77 GB) and identical 4/4 vision. Both versions are now documented as verified, and the stale `# adds mlx-vlm >= 0.6.3` install comment is corrected.

## Declining

**Remove `--quantize-extras` from the full-precision vision build** (README:1500). This conflates two different commands. Line 1500 is the new Qwen3.8 block, which makes no full-precision claim about the vision tower — the card and README both state plainly that embeddings, the vision tower and `lm_head` are 8-bit affine. The "vision towers … stay full precision" comment is at line ~1611, on the pre-existing DiffusionGemma command, which correctly has no `--quantize-extras`.

Removing the flag would also produce a 17.65 GiB build instead of 15.15 GiB, no longer matching the published artifact.

Suite: 551 passed, 5 skipped. README internal anchors verified.

**Comment by manjunathshiva** [comment]
Fixed in a2dbd99 — this one was real, and it was mine.

`type=int` accepted `0`, which **hangs** mlx-vlm's chunked prefill rather than erroring. The loop is:

```python
while inputs_embeds.shape[1] > 1:
    n = min(prefill_step_size, inputs_embeds.shape[1] - 1)
    inputs_embeds = inputs_embeds[:, n:]
```

At step 0, `n` is 0, the slice removes nothing, the length never falls, and it never terminates.

I simulated the real arithmetic rather than guessing at the boundary, which sharpened the claim:

| step | behaviour on a 100-token prompt |
|---|---|
| 1 | terminates in 99 chunks — merely slow, **not** a hang |
| **0** | **no progress: `n_to_process=0` leaves the length unchanged** |
| −1 | `[:, -1:]` slices from the wrong end |

Corroboration that this is a genuine gap and not just defensive coding: mlx-vlm validates exactly this in its **diffusion** path (`prefill_step_size must be a positive integer`, `generate/diffusion.py:655`) but **not** in the autoregressive path that `generate_vlm` uses.

Now an argparse `type=` callable, so it fails as a clean CLI error before anything loads:

```
--prefill-step-size 0   -> error: argument --prefill-step-size: must be a positive integer, got 0
--prefill-step-size -1  -> error: argument --prefill-step-size: must be a positive integer, got -1
--prefill-step-size 256 -> accepted
```

Six tests added covering both directions, and I verified they catch the regression: rebinding `_positive_int` to plain `int` fails all three rejection cases.

Suite: 557 passed, 5 skipped.
